flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/CoalesceTypeStrategy.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.strategies;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.CallContext;
+import org.apache.flink.table.types.inference.TypeStrategy;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeMerging;
+import org.apache.flink.table.types.utils.TypeConversions;
+
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * A type strategy for COALESCE return type. The return type of COALESCE is the least restrictive
+ * type of its arguments, and it's nullable only if all its input args are nullable.
+ */
+@Internal
+public final class CoalesceTypeStrategy implements TypeStrategy {
+
+    public CoalesceTypeStrategy() {}
+
+    @Override
+    public Optional<DataType> inferType(CallContext callContext) {
```

Comment:
This strategy is essentially the same as `CommonTypeStrategy` except for how nullability is determined. We could probably reuse the `CommonTypeStrategy` and instead just create a wrapping strategy that determines the nullability to reduce the duplication here?

Reply:
I've reused `NullableIfArgsTypeStrategy`, removing the code here. Now `TypeStrategies` has both `nullableIfArgs` and `nullableIfAllArgs`. Perhaps should i rename `nullableIfArgs` in `nullableIfAnyArgs`? Should I do it in this PR (separate commit) or another one?